### DISCUSSION TITLE
state with reference types

### DISCUSSION THREAD

**Discussion by JaapWijnen** [pointfreeco/swift-composable-architecture#1157] [Q&A]
The main readme and general usage describes to always use simple value types to build your app state. 
I'm currently working on an app that has a graph like state (reference types) and would like to understand what I should look out for if I'd try to use TCA with that setup and if it's possible.

Part of my state looks somewhat like the following: 

```swift
struct AppState {
  var nodes: [Node]
}

class Node {
  var connections: [NodeConnection]
}

class NodeConnection {
  weak var inputNode: Node?
  weak var outputNode: Node?
}
```
Is it possible to use reference types in TCA state as well? What should I look out for? I'm not sure how to approach it yet.

**Comment by stephencelis** [comment]
In general, I think we would advise against keeping reference types directly in state. It just introduces too many unknowns, like spooky [action at a distance](https://en.wikipedia.org/wiki/Action_at_a_distance_(computer_programming)), and pain points, like manual equality operations that can get out of sync with the stored data, which view stores and SwiftUI heavily rely on for recomputing view state.

Core Data poses a similar problem, and is why those intrepid enough to use TCA and Core Data together typically introduce a translation layer, where they will use the environment to fetch an object and then convert it to a value type before inserting it into state.

For Core Data and other graph databases, the environment seems like a natural place to introduce a client, and then your app's reducers can query this database conditionally depending on the view. Then, any state the view needs can be converted into a value type.

**Reply by JaapWijnen** [reply]
Hmm that's too bad, the reference types are fairly simple. I've been implementing mostly in vanilla SwiftUI at the moment and really miss the benefits of TCA haha! 
I'm basically building a large canvas with draggable nodes on it that can be connected by the user. But there's some state that needs to be shared between siblings (nodes and connections) for positioning to correctly draw them to connect. Technically as they're a graph they're not siblings of course but view wise they are drawn as siblings. 

I definitely miss the ability to repackage/scope TCA state into a computed property to pass to child views. Is there any SwiftUI equivalent to that? 

for example:
```swift
class NodeViewModel {
  var nodeProperty: String
  var connectors: [ConnectorModel]
}

class ConnectorModel {
  var connectorProperty: String 
}

class ConnectorViewModel { // pass this to the view (I need both properties in the viewModel since I need them to compute another value which is used by a sibling which only has access to the connector, not the node)
  var nodeProperty: String
  var connectorProperty: String 
}

struct NodeView: View {
  var body: some View {
    ForEach(connectors) { connector in 
      let connectorViewModel = ... // how to do this 
      ConnectorView(viewModel: connectorViewModel)
    }
  }
}
```
